### PR TITLE
Provide a miragevpn.key binary which generate keys for OpenVPN

### PR DESCRIPTION
**Pull request by dinosaure**
A little draft which provides a tool to generate tls-crypt-v2 key. The goal is to be able to provide all keys needed to instantiate an OpenVPN connection with a nice `man` and arguments.

### CODE REVIEW

**Review by reynir — COMMENTED** [review]
Overall I think this is good. It's unclear to me if the `Tls_crypt_v2` modules is supposed to be used by the rest of the library - in that case I have some wishes.

File: src/dune [hunk 1]
```diff
@@ -9,6 +9,8 @@
   decompress.lzo
   mirage-crypto
   tls
+  hxd.core
+  hxd.string
```
Comment (hannesm) [inline src/dune:13]: do we really want this dependency in the core library?

File: src/tls_crypt_v2.ml [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+  val of_cstruct : Cstruct.t -> (t, [> `Msg of string ]) result
+  val to_cstruct : t -> Cstruct.t
+  val to_string : t -> string
+  val cipher_key : t -> Mirage_crypto.Cipher_block.AES.CTR.key
+  val hmac : t -> Cstruct.t
+  val equal : t -> t -> bool
+  val pp_hum : t Fmt.t
+  val generate : ?g:Mirage_crypto_rng.g -> unit -> t
+  val to_base64 : t -> string
+end = struct
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply src/tls_crypt_v2.ml:61]: It's probably better to have an ADT such as:
```ocaml
type key =
  | Tls_crypt_v2 of Tls_crypt_v2.Key.t
  | Tls_auth _
  | Static _
```

WDYT?

File: src/tls_crypt_v2.ml [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+  val load :
+    key:Server.t -> lines:string Seq.t -> (t, [> `Msg of string ]) result
+
+  val generate :
+    ?g:Mirage_crypto_rng.g -> now:(unit -> Ptime.t) -> Metadata.t option -> t
+
+  val save : key:Server.t -> Client.t -> string Seq.t
+end = struct
+  type t = Key.t * Key.t * Metadata.t
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply src/tls_crypt_v2.ml:191]: Should we define it as an option?

File: src/tls_crypt_v2.ml [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    let* a = Key.of_cstruct (Cstruct.sub kc 0 128) in
+    let* b = Key.of_cstruct (Cstruct.sub kc 128 128) in
+    let* a', b', metadata = unwrap ~key:server_key wkc in
+    let* () =
+      guard ~msg:"Client keys don't correspond" @@ fun () -> Key.equal a a'
+    in
+    let* () =
+      guard ~msg:"Client keys don't correspond" @@ fun () -> Key.equal b b'
+    in
+    Ok (a, b, metadata)
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply src/tls_crypt_v2.ml:270]: I think we need to precise what we want to expose then. The goal of the `unwrap`/`wrap` is to be able to manipulate tls-crypt-v2 keys (and expose a tool) and, may be, have a common part for the server when it receives the `wKc`.

File: src/tls_crypt_v2.ml [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    let* () =
+      guard ~msg:"Can not read tls-crypt-v2 client key length" @@ fun () ->
+      Cstruct.length cs >= 2
+    in
+    let net_len = Cstruct.BE.get_uint16 cs (Cstruct.length cs - 2) in
+    let* () =
+      guard ~msg:"Can not locate tls-crypt-v2 wrapped client key" @@ fun () ->
+      Cstruct.length cs >= net_len
+    in
+    let wkc = Cstruct.sub cs (Cstruct.length cs - net_len) net_len in
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply src/tls_crypt_v2.ml:255]: We probably can improve the API about that.

File: src/tls_crypt_v2.ml [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    let module AES_CTR = Mirage_crypto.Cipher_block.AES.CTR in
+    let ctr = AES_CTR.ctr_of_cstruct tag in
+    let* key_and_metadata =
+      try
+        let payload =
+          Cstruct.sub cs _TLS_CRYPT_V2_TAG_SIZE
+            (net_len - _TLS_CRYPT_V2_TAG_SIZE - 2)
+        in
+        Ok (AES_CTR.decrypt ~key:(Key.cipher_key server_key) ~ctr payload)
+      with _ -> error_msgf "Could not decrypt client key"
```
[the remark this answers is not in the document]
Comment (dinosaure) [reply src/tls_crypt_v2.ml:216]: `net_len` is actually an user input - this one can be wrong.

File: src/dune (hunk 1 above)
Comment (dinosaure) [reply src/dune:13]: Currently, it's used by `pp_hum` of `Tls_crypt_v2.{Client,Server}`. It's not mandatory but the output is nicer than `ocaml-hex` I think. We can also not provide `pp_hum` and think about an other way to print out keys.

File: src/tls_crypt_v2.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (reynir) [reply src/tls_crypt_v2.ml]: static keys, tls-auth and even tls-crypt(-v1) use the same format for keys. You can tell from the man page for `--genkey`:
              secret                Standard OpenVPN shared secret keys
              tls-crypt             Alias for secret
              tls-auth              Alias for secret

Tls-crypt-v2 is only special in that the metadata/wkc is there and the file format is different, but the underlying keys are "the same" (i.e. four times 512 bit key material). I'm not sure what the best way to represent them is. For the library it wouldn't be too useful to use an ADT as you don't want to check for tls-auth keys in Client_tls_crypt_v2. Hmm hmm...

File: src/tls_crypt_v2.ml [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+  val generate :
+    ?g:Mirage_crypto_rng.g -> now:(unit -> Ptime.t) -> Metadata.t option -> t
+
+  val save : key:Server.t -> Client.t -> string Seq.t
+end = struct
+  type t = Key.t * Key.t * Metadata.t
+
+  let server_key (k, _, _) = k
+  let client_key (k, _, _) = k
```
Comment (reynir) [inline src/tls_crypt_v2.ml:194]: I need to look it up, but I think the client key is the second half
```suggestion
  let client_key (_, k, _) = k
```

File: src/tls_crypt_v2.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dinosaure) [reply src/tls_crypt_v2.ml]: You probably want a projection according the key. Like:
```ocaml
type 'a hmac -
   | Tls_crypt_v2 : Mirage_crypto.SHA256.hmac hmac
   | ....

val hmac : t -> 'a hmac -> 'a
```

WDYT?

File: app/key.ml [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    const genkey $ term_random_number_generator $ ty $ metadata $ server_key
+    $ output)
+
+let cmd_genkey =
+  let open Cmdliner in
+  let doc = "A tool to generate tls-crypt-v2 keys." in
+  let man =
+    [
+      `S Manpage.s_description;
+      `P "$(term) is a simple tool to generate tls-crypt-v2 keys.";
```
Comment (reynir) [inline app/key.ml:173]: This prints `undefined is a simple tool[...]`

File: src/tls_crypt_v2.ml [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+  type t = Cstruct.t (* cipher (64 bytes) + hmac (64 bytes) *)
+
+  let len_of_secret_aes_ctr_key =
+    Array.to_list Mirage_crypto.Cipher_block.AES.CTR.key_sizes
+    |> List.filter ((>) 128) (* ∀x. 128 > x <=> ∀x. x <= 128 *)
+    |> List.fold_left max 0
+
+  (* NOTE: it's a bit paranoid but we ensure that [mirage-crypto] exposes values
+     which fit with an OpenVPN key (128 bytes). *)
+  let () = assert (len_of_secret_aes_ctr_key > 0)
```
Comment (reynir) [inline src/tls_crypt_v2.ml:71]: I would go with `((>=) 64)` as each key (128 bytes) is a pair of cipher key material (up to 64 bytes) and hmac key material (up to 64 bytes).

File: src/dune (hunk 1 above)
Comment (reynir) [reply src/dune]: I moved the `Tls_crypt_v2.Key.pp_hum` function into the application so the library can drop the hxd dependency. It is not used anywhere though.

**Comment by reynir** [comment]
Let's merge this and then take a look at using the key modules in the library as well at a later point.

**Pull request merged**